### ISSUE TITLE
Method to get maximum number of bytes necessary for match

### ISSUE THREAD

**Issue by vlmutolo** [rust-lang/regex#725]
TL;DR: Is there a way to ask a `Regex` struct the number of bytes it needs to see to find a match? Basically I’m interested in streaming matches. 

I’m a little fuzzy on exactly what I’m thinking of for this, so I’ll start with my intended use case to avoid an XY problem. 

I’m experimenting with extending the `im` crate to provide an immutable string based on their `Vector` type. I’m thinking through the various uses of std’s `String` and trying to port them to my new type. I started wondering how to do searches on this new kind of string, which is essentially a discontiguous rope, and realized it’s kind of nontrivial to do a regex search because matches might be found on the boundary between two contiguous chunks of the full string. 

So I thought it could be interesting to provide an interface to allocate a buffer of some kind that can store at least the necessary number of bytes for a regex match. 

For example, for the regex `(hi|hello)`, I figure you never need to look at more than five bytes at a time to determine if there’s a match. So I can just search contiguous strings like normal, then load the last five bytes of the str and the first five of the next into the buffer, search that like normal, and then continue into the next contiguous str like normal. 

I haven’t worked out all the details yet, but the general strategy relies on being able to ask the `Regex` type how many bytes I need to see at a time. Do you know if there’s a way to do that, or if not, how hard such an API addition would be?

**Comment by ethanpailes** [comment]
@vlmutolo this crate does not support streaming regex. I know some other folks have been interested in this sort of thing in the past, so one of them may have made some progress on it. You can search through the issue tracker to look for conversations about it.

To see why what you are asking for is tricky, consider the regex `a*`. How many bytes must we allocate in order to match that regex? We can't know without looking at the actual input data, in which case we might as well just perform the match on the streaming input directly.

The matchers in this crate are pretty tightly coupled to the representation of an input string as a slice of bytes for performance reasons. If you fork this crate and disable all of the matchers except the PikeVM (that matcher is pretty simple), then update that matcher to operate on a streaming API, you may have some luck with this.

**Comment by vlmutolo** [comment]
Yeah I can see why it would be a difficult problem. For my use case, I can probably settle for implementing some kind of `lfind` and `rfind` like std does. Anyway, thanks for the quick response!

**Comment by BurntSushi** [comment]
This issue documents a lot of thinking and possible avenues forward with respect to stream or discontiguous matching: https://github.com/rust-lang/regex/issues/425

Some short notes:

* The `regex-automata` crate exposes a low level interface to a traditional DFA and [folks have used this interface to implement a form of stream matching](https://github.com/rust-lang/regex/issues/425#issuecomment-706804461).
* The [`aho-corasick` crate supports a limited form of stream searching](https://docs.rs/aho-corasick/0.7.15/aho_corasick/struct.AhoCorasick.html#method.stream_find_iter). I'd encourage you to look at how complex even a simple case like this is to implement. Or at least, if there is a simpler approach, I'd love to learn about it.
